log/format.go

```diff
@@ -156,55 +161,74 @@ func TerminalFormat(usecolor bool) Format {
 //
 // For more details see: http://godoc.org/github.com/kr/logfmt
 func LogfmtFormat() Format {
-	return FormatFunc(func(r *Record) []byte {
-		common := []interface{}{r.KeyNames.Time, r.Time, r.KeyNames.Lvl, r.Lvl, r.KeyNames.Msg, r.Msg}
+	return FormatFunc(func(r slog.Record) []byte {
+		common := []slog.Attr{slog.Time(timeKey, r.Time), slog.Int(lvlKey, int(r.Level)), slog.String(msgKey, r.Message)}
 		buf := &bytes.Buffer{}
-		logfmt(buf, append(common, r.Ctx...), 0, false)
+		logfmt(buf, common, r, 0, false)
 		return buf.Bytes()
 	})
 }
 
-func logfmt(buf *bytes.Buffer, ctx []interface{}, color int, term bool) {
-	for i := 0; i < len(ctx); i += 2 {
+func evaluateIfLazy(val interface{}) interface{} {
+	if lz, ok := val.(Lazy); ok {
+		rawVal, err := evaluateLazy(lz)
+		if err != nil {
+			return err
+		}
+
+		return rawVal
+	}
+	return val
+}
+
+func logfmt(buf *bytes.Buffer, commonAttrs []slog.Attr, r slog.Record, color int, term bool) {
+	attrs := []slog.Attr{}
+	r.Attrs(func(attr slog.Attr) bool {
+		attrs = append(attrs, attr)
+		return true
+	})
+
+	attrs = append(commonAttrs, attrs...)
+
+	for i, attr := range attrs {
 		if i != 0 {
 			buf.WriteByte(' ')
 		}
 
-		k, ok := ctx[i].(string)
-		v := formatLogfmtValue(ctx[i+1], term)
-		if !ok {
-			k, v = errorKey, fmt.Sprintf("%+T is not a string key", ctx[i])
-		} else {
-			k = escapeString(k)
-		}
+		key := escapeString(attr.Key)
+		rawVal := evaluateIfLazy(attr.Value.Any())
+		val := formatLogfmtValue(rawVal, true)
 
 		// XXX: we should probably check that all of your key bytes aren't invalid
+		// TODO (jwasinger) above comment was from log15 code.  what does it mean?  check that key bytes are ascii characters?
 		fieldPaddingLock.RLock()
-		padding := fieldPadding[k]
+		padding := fieldPadding[key]
 		fieldPaddingLock.RUnlock()
 
-		length := utf8.RuneCountInString(v)
+		length := utf8.RuneCountInString(val)
 		if padding < length && length <= termCtxMaxPadding {
 			padding = length
 
 			fieldPaddingLock.Lock()
-			fieldPadding[k] = padding
+			fieldPadding[key] = padding
 			fieldPaddingLock.Unlock()
 		}
 		if color > 0 {
-			fmt.Fprintf(buf, "\x1b[%dm%s\x1b[0m=", color, k)
+			fmt.Fprintf(buf, "\x1b[%dm%s\x1b[0m=", color, key)
 		} else {
-			buf.WriteString(k)
+			buf.WriteString(key)
 			buf.WriteByte('=')
 		}
-		buf.WriteString(v)
-		if i < len(ctx)-2 && padding > length {
+		buf.WriteString(val)
+		if i < r.NumAttrs()-1 && padding > length {
 			buf.Write(bytes.Repeat([]byte{' '}, padding-length))
 		}
 	}
 	buf.WriteByte('\n')
 }
 
+// TODO: investigate use of built-in slog json handler so we can remove the one here
```

Comment:
Yes, definitely seems like something that `slog` should support out of the box

Reply:
It does.  Thinking about it more, it actually should be straightforward to use slog's built-in handlers for logfmt/json formats and keep the existing logfmt in the Geth codebase.  These can be sub-handlers.